### ISSUE TITLE
False positive for FL_FLOATS_AS_LOOP_COUNTERS

### ISSUE THREAD

**Issue by DanySK** [spotbugs/spotbugs#2126]
Relevant code:
```java
double y = Double.NaN
while (Double.isNaN(y) && someOtherCondition) {
    ...
    y = Double.parseDouble(...);
    ...
}
```

Reported:

#### FL_FLOATS_AS_LOOP_COUNTERS: Do not use floating-point variables as loop counters

Using floating-point variables should not be used as loop counters, as they are not precise, which may result in incorrect loops. A loop counter is a variable that is changed with each iteration and controls when the loop should terminate. *It is decreased or increased by a fixed amount each iteration*.

See rule [NUM09-J](https://wiki.sei.cmu.edu/confluence/display/java/NUM09-J.+Do+not+use+floating-point+variables+as+loop+counters).

---

However, there is no fixed increment, so I think it does not apply here.

Bytecode:
[FollowTarget.zip](https://github.com/spotbugs/spotbugs/files/9226167/FollowTarget.zip)

**Comment by DanySK** [comment]
Another false positive here:
```java
    protected static WeibullDistribution weibullFromMean(
            final double mean,
            final double deviation,
            final RandomGenerator random
    ) {
        final double t = FastMath.log((deviation * deviation) / (mean * mean) + 1);
        double kmin = 0, kmax = 1;
        while (Gamma.logGamma(1 + 2 * kmax) - 2 * Gamma.logGamma(1 + kmax) < t) {
            kmin = kmax;
            kmax *= 2;
        }
        double k = (kmin + kmax) / 2;
        while (kmin < k && k < kmax) { // WARNING
            if (Gamma.logGamma(1 + 2 * k) - 2 * Gamma.logGamma(1 + k) < t) {
                kmin = k;
            } else {
                kmax = k;
            }
            k = (kmin + kmax) / 2;
        }
        final double shapeParameter = 1 / k;
        final double scaleParameter = mean / FastMath.exp(Gamma.logGamma(1 + k));
        return new WeibullDistribution(
                random,
                shapeParameter,
                scaleParameter,
                WeibullDistribution.DEFAULT_INVERSE_ABSOLUTE_ACCURACY
        );
    }
```

**Comment by atnak** [comment]
Here's another vastly simplified false positive:
```java
double total = 0;

for (int attempt = 1;; attempt++) {
	total += RandomUtils.nextDouble(); // FL_FLOATS_AS_LOOP_COUNTERS

	if (RandomUtils.nextDouble() < 0.5) {
		System.out.println(attempt);
		break;
	}
}
```

Two things that matter seem to be:
1. Having something (anything) in the *increment* part of the for-statement
2. Having a assignment to a floating point variable as the first statement the loop

**Comment by KengoTODA** [comment]
closed by #2167

**Comment by bradh** [comment]
I'm still hitting this with 4.7.2, withhttps://github.com/WestRidgeSystems/jmisb/blob/2.x/api-ffmpeg/src/main/java/org/jmisb/api/video/VideoFileInput.java#L286-L288

I'm having a hard time making a smaller example than that though. Even extracting the whole module doesn't reproduce it.
